torch/csrc/jit/passes/xnnpack_rewrite.cpp

```diff
@@ -0,0 +1,121 @@
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/xnnpack_rewrite.h>
+#include <torch/csrc/jit/passes/fuse_linear.h>
+#include <torch/csrc/jit/passes/graph_rewrite_helper.h>
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
+#include <torch/csrc/jit/ir/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+namespace {
+std::string getFuncName(Value* func_value) {
+  auto func_node = func_value->node();
+  auto func = func_node->output()->type()->expect<FunctionType>()->function();
+  const auto& qname = func->qualname();
+  const auto& name = qname.qualifiedName();
+  auto rdot_idx = name.rfind('.');
+  if (rdot_idx != std::string::npos) {
+    return name.substr(rdot_idx + 1, name.length());
+  } else {
+    return name;
+  }
+}
+
+void insertXNNPACKLinearOp(std::shared_ptr<Graph>& graph) {
+  std::string linear_before_inline = R"(
+    graph(%linear, %input, %weight, %bias):
+        %r = prim::CallFunction(%linear, %input, %weight, %bias)
+        return (%r))";
+  std::string xnnpack_pattern_before_inline = R"(
+    graph(%linear, %input, %weight, %bias):
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
+  std::string linear_pattern = R"(
+    graph(%input, %weight, %bias):
+        %r = aten::linear(%input, %weight, %bias)
+        return (%r))";
+  std::string xnnpack_pattern = R"(
+    graph(%input, %weight, %bias):
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
+  std::string linear_pattern_wo_bias = R"(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
+        %res = aten::linear(%input, %weight, %bias)
+        return (%res))";
+  std::string xnnpack_pattern_wo_bias = R"(
+    graph(%input, %weight):
+        %bias: Tensor? = prim::Constant()
+        %packed_weight_bias = xnnpack::linear_prepack(%weight, %bias)
+        %res = xnnpack::linear_packed(%input, %packed_weight_bias)
+        return (%res))";
```

Comment:
This is probably not needed

Reply:
Why not?

Reply:
this is the same as previous pattern

Reply:
Previous pattern is supposed to catch bias=None as well? I will try it out.